src/lex/cpp.rs

```diff
@@ -1161,16 +1161,16 @@ impl<'a> PreProcessor<'a> {
     fn update_builtin_definitions(&mut self) {
         self.definitions.extend(map! {
             "__LINE__".into() => int_def((self.line() + 1) as i32),
-            "__FILE__".into() => str_def(self.file_processor.path().to_str().unwrap()),  // Fails if non-Unicode
+            "__FILE__".into() => str_def(self.file_processor.path().to_string_lossy()),
         })
     }
 }
 
 fn int_def(i: i32) -> Definition {
     Definition::Object(vec![LiteralToken::Int(RcStr::from(i.to_string())).into()])
 }
-fn str_def(s: &str) -> Definition {
-    let rcstr = RcStr::from(format!("\"{}\"", s));
+fn str_def<S: ToString>(s: S) -> Definition {
+    let rcstr = RcStr::from(format!("\"{}\"", s.to_string().replace(r#"""#, r#"\""#)));
```

Comment:
Use `Into` instead; `ToString` is meant for displaying things to the end user and is more flexible than we need here (it includes things like numbers and ip addresses): https://doc.rust-lang.org/std/string/trait.ToString.html

```suggestion
fn str_def<S: Into<String>>(s: S) -> Definition {
    let rcstr = RcStr::from(format!("\"{}\"", s.into().replace(r#"""#, r#"\""#)));
```

Reply:
Good to know, thanks!